sql/core/src/test/scala/org/apache/spark/sql/crossdbms/CrossDbmsQueryTestSuite.scala

```diff
@@ -0,0 +1,278 @@
+ * {{{
+ *   bash ./bin/generate_golden_files_with_postgres.sh
+ * }}}
+ * 2.
+ *   a. You need to have a Postgres server up before running this test.
+ *      i. Install PostgreSQL. On a mac: `brew install postgresql@13`
+ *      ii. After installing PostgreSQL, start the database server, then create a role named pg with
+ *      superuser permissions: `createuser -s postgres` OR `psql> CREATE role postgres superuser`
+ *   b. Run the following command:
+ *     {{{
+ *       SPARK_GENERATE_GOLDEN_FILES=1 build/sbt "sql/testOnly org.apache.spark.sql.crossdbms.PostgreSQLQueryTestSuite"
+ *     }}}
+ *
+ * To indicate that the SQL file is not eligible for testing with this suite, add the following
+ * comment into the input file:
+ * {{{
+ *   --ONLY_IF spark
+ * }}}
+ *
+ * And then, to run the entire test suite, with the default cross DBMS:
+ * {{{
+ *   build/sbt "sql/testOnly org.apache.spark.sql.crossdbms.PostgreSQLQueryTestSuite"
+ * }}}
+ *
+ * To re-generate golden file for a single test, e.g. `describe.sql`, run:
+ * {{{
+ *   SPARK_GENERATE_GOLDEN_FILES=1 build/sbt "sql/testOnly org.apache.spark.sql.crossdbms.PostgreSQLQueryTestSuite -- -z describe.sql"
+ * }}}
+ */
+
+class PostgreSQLQueryTestSuite extends CrossDbmsQueryTestSuite {
+
+  protected def crossDbmsToGenerateGoldenFiles: String = CrossDbmsQueryTestSuite.POSTGRES
+
+  // Reduce scope to subquery tests for now. That is where most correctness issues are.
+  override protected def customInputFilePath: String = new File(inputFilePath, "subquery").getAbsolutePath
+
+  override protected def getConnection: Option[String] => JdbcSQLQueryTestRunner =
+    (connection_url: Option[String]) => JdbcSQLQueryTestRunner(PostgresConnection(connection_url))
+
+  override protected def preprocessingCommands = Seq(
+    // Custom function `double` to imitate Spark's function, so that more tests are covered.
+    """
+      |CREATE OR REPLACE FUNCTION double(numeric_value numeric) RETURNS double precision
+      |    AS 'select CAST($1 AS double precision);'
+      |    LANGUAGE SQL
+      |    IMMUTABLE
+      |    RETURNS NULL ON NULL INPUT;
+      |""".stripMargin
+  )
+}
+
+/**
+ * See SQLQueryTestSuite.scala for more information. This suite builds off of that to allow us
+ * to generate golden files with other DBMS to perform cross-checking for correctness. It generates
+ * another set of golden files. Note that this is not currently run on all SQL input files by
+ * default because there is incompatibility between SQL dialects for Spark and the other DBMS.
+ *
+ * This suite adds a new comment argument, --ONLY_IF. This comment is used to indicate the DBMS for
+ * which is eligible for the SQL file. These strings are defined in the companion object. For
+ * example, if you have a SQL file named `describe.sql`, and you want to indicate that Postgres is
+ * incompatible, add the following comment into the input file:
+ * --ONLY_IF spark
```

Comment:
This could be simpler by just inferring whether to test against e.g. Postgres or not based on the presence of the corresponding golden file in its associated directory instead.

Reply:
I wanted to do this so that new sql test files are automatically opted into testing against Postgres. This "forces" the developer who adds the new test files to either generate golden files with Postgres, or actively ignore it.